lxd/storage/drivers/driver_powerflex_utils.go

```diff
@@ -0,0 +1,1021 @@
+		if err != nil {
+			return fmt.Errorf("Failed to read response body: %s: %w", path, err)
+		}
+	}
+
+	// Return the formatted error from the body
+	powerFlexErr, ok := v.(*powerFlexError)
+	if ok {
+		return powerFlexErr
+	}
+
+	return nil
+}
+
+// requestAuthenticated issues an authenticated HTTP request against the PowerFlex gateway.
+func (p *powerFlexClient) requestAuthenticated(method string, path string, body io.Reader, v any) error {
+	err := p.login()
+	if err != nil {
+		return err
+	}
+
+	return p.request(method, path, body, v)
+}
+
+// login creates a new access token and authenticates the client.
+func (p *powerFlexClient) login() error {
+	if p.token.Get() != "" {
+		return nil
+	}
+
+	body, err := p.createBodyReader(map[string]any{
+		"username": p.driver.config["powerflex.user.name"],
+		"password": p.driver.config["powerflex.user.password"],
+	})
+	if err != nil {
+		return err
+	}
+
+	var actualResponse struct {
+		AccessToken string        `json:"access_token"`
+		ExpiresIn   time.Duration `json:"expires_in"`
+	}
+
+	err = p.request(http.MethodPost, "/rest/auth/login", body, &actualResponse)
+	if err != nil {
+		return fmt.Errorf("Failed to login: %w", err)
+	}
+
+	// Expire the token after half of its lifetime
+	p.token.Set(actualResponse.AccessToken, actualResponse.ExpiresIn*time.Second/2)
+	return nil
+}
+
+// getStoragePool returns the storage pool behind poolID.
+func (p *powerFlexClient) getStoragePool(poolID string) (*powerFlexStoragePool, error) {
+	var actualResponse powerFlexStoragePool
+	err := p.requestAuthenticated(http.MethodGet, fmt.Sprintf("/api/instances/StoragePool::%s", poolID), nil, &actualResponse)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to get storage pool: %q: %w", poolID, err)
+	}
+
+	return &actualResponse, nil
+}
+
+// getProtectionDomain returns the protection domain behind domainID.
+func (p *powerFlexClient) getProtectionDomain(domainID string) (*powerFlexProtectionDomain, error) {
+	var actualResponse powerFlexProtectionDomain
+	err := p.requestAuthenticated(http.MethodGet, fmt.Sprintf("/api/instances/ProtectionDomain::%s", domainID), nil, &actualResponse)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to get protection domain: %q: %w", domainID, err)
+	}
+
+	return &actualResponse, nil
+}
+
+// getProtectionDomainSdsRelations returns the protection domains SDS relations.
+func (p *powerFlexClient) getProtectionDomainSdsRelations(domainID string) ([]powerFlexProtectionDomainSdsRelation, error) {
+	var actualResponse []powerFlexProtectionDomainSdsRelation
+	err := p.requestAuthenticated(http.MethodGet, fmt.Sprintf("/api/instances/ProtectionDomain::%s/relationships/Sds", domainID), nil, &actualResponse)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to get protection domain SDS relations: %q: %w", domainID, err)
+	}
+
+	return actualResponse, nil
+}
+
+// getVolumeID returns the volume ID for the given name.
+func (p *powerFlexClient) getVolumeID(name string) (string, error) {
```

Comment:
Please can you avoid the use of generic `name` args, and make them a little more verbose, e.g. `volumeName` or `volName` for the avoidance of doubt as to what name is.

Reply:
Fixed.